src/librustc_typeck/collect/type_of.rs

```diff
@@ -403,7 +403,9 @@ fn find_opaque_ty_constraints(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                 for (i, arg) in substs.iter().enumerate() {
                     let arg_is_param = match arg.unpack() {
                         GenericArgKind::Type(ty) => matches!(ty.kind, ty::Param(_)),
-                        GenericArgKind::Lifetime(lt) => !matches!(lt, ty::ReStatic),
+                        GenericArgKind::Lifetime(lt) => {
+                            matches!(lt, ty::ReEarlyBound(_) | ty::ReFree(_))
+                        }
```

Comment:
I guess the other types just didn't happen in practice? Else it'd be nice to have a test.

Reply:
These are "exported" by the MIR borrowck (cc @matthewjasper) AFAICT, and that results in late-bound lifetimes (in the `fn` signature, I presume) being `ReFree`.

This list is minimal, as in I can't remove either `ReEarlyBound` nor `ReFree` without tests failing, and adding more would cause less errors (i.e. allowing buggy situations), not more.

Reply:
Actually, @matthewjasper, shouldn't `ReFree` turn back into `ReLateBound`? Although, I guess the actual concrete type is in terms of the `type Foo<...> = impl Trait;` generics, so it's not affected by the substitutions back at the point `Foo` is mentioned.

Reply:
I don't think that we currently ever go from a placeholder back to a bound region/type.